File: Esay/Two Sum.cpp
//Question link:-https://www.naukri.com/code360/problems/two-sum_839653
#include <iostream>
#include <vector>
#include <unordered_map>
#include <set>
#include <utility> 
#include <bits/stdc++.h>
vector<pair<int,int>> twoSum(vector<int>& arr, int target, int n)
{
  sort(arr.begin(), arr.end());
  vector<pair<int, int>> ans;

  int i = 0, j = n - 1;
  int sum = 0;
  bool found = false;

  while (i < j) {
    sum = arr[i] + arr[j];
    if (sum < target)
      i++;
    else if (sum > target)
      j--;
    else if (sum == target) {
      ans.push_back(make_pair(arr[i], arr[j]));
	  i++;
	  j--;
	  found = true;
    }
  }

  if(found == false)
  	ans.push_back(make_pair(-1, -1));

  return ans;
}

File: Easy/two-sum.java
//https://leetcode.com/problems/two-sum/description/
class Solution {
    public int[] twoSum(int[] nums, int target) {
        //Arrays.sort(nums);
        int[] arr= new int[2];
        // for(int i=0;i<nums.length;i++){          //O(n^2)
        //     int Newtarget=target-nums[i];
        //     for(int j=i+1;j<nums.length;j++){
        //         if(nums[j]==Newtarget){
        //             arr[0]=i;
        //             arr[1]=j;
        //             return arr;
        //         }
        //     }
        // }
        Map<Integer, Integer> map = new HashMap<>();    //O(n)
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            map.put(nums[i], i);
        }
        return arr;
    }
}
